### PR TITLE
RFC: Different approach to Editor tutorial (see #189)

### PR DESCRIPTION
**Pull request by MatthiasWM**
tl;dr: do you like the tutorial approach for 'editor' on teaching FLTK ? If yes, I will wrap this up for 1.4.0 by Sunday Jan 7th.

https://github.com/MatthiasWM/fltk/blob/editor2/documentation/src/editor.dox
https://github.com/MatthiasWM/fltk/blob/editor2/test/editor.cxx

Before I continue here and you guys don't like my approach:

I started (and half-way finished) to rewrite the documentation for the editor section. I tried to write the code in a way that readers can follow along. Each chapter by itself creates working code which can be activated by setting a macro to the chapter number. Each subsequent chapter adds new aspects to the previous code.

I kept the original idea of "editor", but I tried to be more explicit in functions names, trying to make more obvious what we are doing here (for example, end functions in `_callback` instead of `_cb`). The resulting editor will be very similar. I did change the concept of multiple windows editing the same file into one window that can have a split editor for the same file. I thing that this better suits this great feature. Also, I will move the support of syntax highlighting to the very end of the source code, making it much less dominant if users do their first small steps in FLTK programming.

Please let me know if you want me to continue, and I will wrap this up tomorrow (Jan 7th), so we can still throw this into 1.4.

### CODE REVIEW

File: documentation/src/editor.dox [hunk 1]
```diff
@@ -3,23 +3,345 @@
...
+
+Phew. I think we put a pretty nice text editor app together in less than
+100 lines of code. The following chapter add some user convenience and show
+off some FLTK features including split editors and syntax highlighting.
+
+
+\section editor_cut_copy_paste Chapter 5: Cut, Copy, and Paste
+
+The FLTK Text Editor widget comes with builtin cut, copy, and
+paste functionality, but as a curtesy, we should also offer these
```
Comment (Albrecht-S) [inline documentation/src/editor.dox:316]: courtesy

**Comment by Albrecht-S** [comment]
+1 :+1: I like the idea, please continue.

I didn't do a full review yet because I wanted to wait until we (you) decided to continue this effort.

One note though: I don't like such personal statements in our documentation like "**_I like to do_** _that by writing a minimal program_..." (after years nobody knows who "I" was). I'd prefer "It is a good idea ..." or "It's a good starting point..." or something like that.

I would also like to add a minimal `CMakeLists.txt` [1] for users wanting to build the project themselves, and I would propose to ignore `configure` and `Makefiles` altogether in this tutorial. This would be good for FLTK 1.4.0 and even more appropriate for 1.5 when we remove configure support.

----------------------
[1] I can contribute this as soon as I pushed the CMake updates I'm working on now.

File: documentation/src/editor.dox [hunk 2]
```diff
@@ -3,23 +3,345 @@
...
+
+The text modify callback sets our `text_changed` flag:
+
+\code
+void text_changed_callback(int, int, int, int, const char*, void*) {
+  text_changed = true;
+}
+\endcode
+
+Lastly, I added a "File/New" menu item to the main menu and gave
```
Comment (Albrecht-S) [inline documentation/src/editor.dox:171]: Suggested change: "Lastly, we add a "File/New" menu item to the main menu and give"

Note: "we" and present tense (my personal opinion).

File: documentation/src/editor.dox [hunk 3]
```diff
@@ -3,23 +3,345 @@
...
+    app_text_buffer->savefile(file_chooser.filename());
+    strncpy(app_filename, file_chooser.filename(), FL_PATH_MAX-1);
+    text_changed = false;
+  }
+}
+\endcode
+
+However if the user has already set a file name including path information,
+it is the polite thing to preload the file chooser with that information. This
+litte chunk of code will separate the file name from the path before we call
```
Comment (Albrecht-S) [inline documentation/src/editor.dox:218]: little

File: documentation/src/editor.dox
Comment (ManoloFLTK) [inline documentation/src/editor.dox:1]: Very nice idea! Yes, please, continue and commit this tutorial.

I also support Albrecht's guidance to refrain from using "I".

**Comment by MatthiasWM** [comment]
Ok, thanks for the positive response. I completed the text, but haven't pro-read it yet. I removed all "I" references. The syntax highlighting is somewhat broken ATM. No clue what I changed ;-).

**Comment by Albrecht-S** [comment]
Please let us know when it's time for a full review (and test), including typo fixes. I'm good at finding typos!  ;-)

**Comment by MatthiasWM** [comment]
I purposely removed all typos... . Sorry, Albrecht ;-)

**Comment by Albrecht-S** [comment]
Nope, sorry, Matthias. See for instance notes above:
- s/litte/little/
- s/curtesy/courtesy/

... unless GitHub doesn't show me the correct diff. I didn't look for more (yet) ...

**Comment by MatthiasWM** [comment]
> Nope, sorry, Matthias. See for instance notes above:
> 
> * s/litte/little/
> * s/curtesy/courtesy/

Aw man! I read it three times and I did not see them. ;-)

**Comment by Albrecht-S** [comment]
@MatthiasWM I'm now reviewing the docs and the new editor.cxx code. Thanks for this big improvement so far. I'll let you know what I found out later.

If you don't mind I can commit (only simple) fixes to your branch. Would you like me to do this rather than sending patches?

One more question: You wrote:
> better know as the cauliflower, or simply "the Apple key".

I thought this key way commonly known as the `cloverleaf` key rather than `cauliflower` (I never read that name for this key before but now I found it mentioned elsewhere). Shouldn't we keep it consistent with other mentions (at least once) and call it `cloverleaf` key?

And BTW, IMHO this should read "better known" (with 'n').

Another long standing issue with our text editor app is IMHO that we should have an option to switch to a fixed font, or even better start with a fixed font and to toggle fixed/proportional font (no complicated font selection necessary). Since you added good looking menu options anyway, wouldn't this be good to add now rather than later?

**Comment by Albrecht-S** [comment]
A first and preliminary patch:
```C++
diff --git a/test/editor.cxx b/test/editor.cxx
index 3fd253369..7dcfdd516 100644
--- a/test/editor.cxx
+++ b/test/editor.cxx
@@ -1,7 +1,7 @@
 //
 // A simple text editor program for the Fast Light Tool Kit (FLTK).
 //
-// This program is described in Chapter 4 of the FLTK Programmer's Guide.
+// This program is described in Chapter 1.7 of the FLTK Programmer's Guide.
 //
 // Copyright 1998-2024 by Bill Spitzak and others.
 //
@@ -116,6 +116,7 @@ void tut3_build_main_editor() {
   app_editor = new Fl_Text_Editor(0, app_menu_bar->h(),
     app_window->w(), app_window->h() - app_menu_bar->h());
   app_editor->buffer(app_text_buffer);
+  app_editor->textfont(FL_COURIER);
   app_window->resizable(app_editor);
   app_window->end();
   // find the Quit menu and insert the New menu there
```
1. Chapter number taken from PDF docs (HTML docs don't seem to show "chapter numbers".
2. Use FL_COURIER as default font, i.e. when syntax highlighting is off.

**Comment by MatthiasWM** [comment]
Hi @Albrecht-S . Feel free to change whatever you see is wrong or questionable. `cloverleaf` is of course correct, but I guess I had brussels sprouts that day and got confused ;-) . 

I can surely add subchapters about font settings too. It's a bit tricky to keep things "sequential", but I will try.

**Comment by erco77** [comment]
Since this is an RFC:

- I try to avoid referring to things like chapter or page numbers in documentation, in favor of using doxygen `\ref` names so that they turn into links in both html and PDF docs. (I do this in for LibreOffice word documentation manuals too; Chapter/section numbers are so 20th century :^D

- If you ever find yourself needing to check documentation, you can use this free grammar/spell checking tool by pasting a few paragraphs at a time into it: https://languagetool.org/

**Comment by Albrecht-S** [comment]
@MatthiasWM I'm working through the chapters and I'm now at chapter 6 of 10. I have some more fixes (typos etc.) and I can commit it altogether if you agree to set the `FL_COURIER` font as the default text font. I tested this and AFAICT it's sufficient to set it in chapter 3 once (I modified my patch to show this in the docs as well).

[not posting the modified patch intentionally for brevity]

**Comment by Albrecht-S** [comment]
@erco77 Thanks for your comment.
> I try to avoid referring to things like chapter or page numbers in documentation, in favor of using doxygen `\ref` names so that they turn into links in both html and PDF docs.

The reference to the chapter number is in `test/editor.cxx` in a normal comment (not doxygen docs). However, I also thought of using the chapter **_title_** rather than the chapter **_number_**.

I changed this in my local copy (to be committed to Matt's branch soon - after dinner ;-) ).

**Comment by Albrecht-S** [comment]
@MatthiasWM OK, I'm done with my review. I pushed commit 231a4e36405e31b23b98ac811230e26fdb425d0a with all my changes to your branch. Please feel free to squash and merge (no need to keep my commit message).

I hope I got everything right, particularly the duplicate stuff in test/editor.cxx and the docs. From my POV this is ready to merge.

BTW: good code and docs, thanks very much. I like it and I assume that users can follow the tutorial. Good job.

**Comment by Albrecht-S** [comment]
Sorry, I forgot the "cauliflower" update, now done in commit 0de5dccc034f3515375486ba588327e14d989620.

**Review by Albrecht-S — APPROVED** [review]
Everything looks good now.

**Comment by MatthiasWM** [comment]
Thanks for fixing all the errors!
